### PR TITLE
fix(grc721): migrate royalty to basis points per EIP-2981

### PR DESCRIPTION
**Pull request by notJoon**
## Summary

related to #1832

This PR migrates the royalty calculation in `p/demo/tokens/grc721` to use basis points (bps, denominator = 10,000) in accordance with the [EIP-2981](https://eips.ethereum.org/EIPS/eip-2981) specification.

The previous integer percentage approach (denominator = 100) cannot represent fractional royalties such as `2.5%` or `7.5%`.

EIP-2981 defines the royalty calculation as:

```
royaltyAmount = salePrice * feeNumerator / feeDenominator
```

In practice, `_feeDenominator` is typically set to `10,000`.

For example, in OpenZeppelin’s ERC2981 implementation, you can see that BPS is set to 10,000:

[https://github.com/OpenZeppelin/openzeppelin-contracts/blob/9cfdccd35350f7bcc585cf2ede08cd04e7f0ec10/contracts/token/common/ERC2981.sol#L75-L82](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/9cfdccd35350f7bcc585cf2ede08cd04e7f0ec10/contracts/token/common/ERC2981.sol#L75-L82)

Regarding naming, we considered keeping `Percent` since the `RoyalInfo` type might be used externally. However, since it is semantically inaccurate, we decided to rename it to `Bps`.

Additionally, to better align with the referenced implementation, renaming it to `FeeDenominator` could also be a good option.

The `calculateRoyaltyAmount` helper function previously returned an error, but since the value was effectively meaningless, it has been inlined and the related error constants have been removed.

### CODE REVIEW

**Comment by Gno2D2** [comment]
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: notJoon/gno-core)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🟢 User davd-gzl already reviewed PR 5504 with state APPROVED
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

File: examples/gno.land/p/demo/tokens/grc721/grc721_royalty_test.gno [hunk 1]
```diff
@@ -26,38 +26,38 @@ func TestSetTokenRoyalty(t *testing.T) {
 
 	derr := dummy.SetTokenRoyalty(TokenID("1"), RoyaltyInfo{
 		PaymentAddress: paymentAddress,
-		Percentage:     percentage,
+		Bps:            bps,
 	})
 	uassert.NoError(t, derr, "Should not result in error")
 
 	// Test case: Invalid token ID
 	_ = dummy.SetTokenRoyalty(TokenID("3"), RoyaltyInfo{
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/tokens/grc721/grc721_royalty_test.gno]:
```suggestion
	derr = dummy.SetTokenRoyalty(TokenID("3"), RoyaltyInfo{
```

File: examples/gno.land/p/demo/tokens/grc721/grc721_royalty.gno [hunk 2]
```diff
@@ -68,12 +71,7 @@ func (r *royaltyNFT) RoyaltyInfo(tid TokenID, salePrice int64) (address, int64,
 	royaltyInfo := val.(RoyaltyInfo)
 
 	// Calculate royalty amount
-	royaltyAmount, _ := r.calculateRoyaltyAmount(salePrice, royaltyInfo.Percentage)
+	royaltyAmount := overflow.Mul64p(salePrice, royaltyInfo.Bps) / bpsDenominator
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/tokens/grc721/grc721_royalty.gno]: If `salePrice` is too high (possible situation) this can overflow.

Consider:
```go
product, ok := overflow.Mul64(salePrice, royaltyInfo.Bps)
if ok {
    royaltyAmount = product / bpsDenominator
} else {
    royaltyAmount = (salePrice / bpsDenominator) * royaltyInfo.Bps
}
```

(division first lose information due to float nature, so it cannot be applied everytime)

Comment (davd-gzl) [reply examples/gno.land/p/demo/tokens/grc721/grc721_royalty.gno]: And you should add a test for this overflow

File: examples/gno.land/p/demo/tokens/grc721/grc721_royalty_test.gno [hunk 3]
```diff
@@ -67,3 +67,38 @@ func TestSetTokenRoyalty(t *testing.T) {
...
+		name       string
+		bps        int64
+		salePrice  int64
+		wantAmount int64
+	}{
+		{"2.5% on 1000", 250, 1000, 25},
+		{"0.5% on 1000", 50, 1000, 5},
+		{"7.5% on 1000", 750, 1000, 75},
+		{"10% on 1000", 1000, 1000, 100},
+		{"100% on 1000", 10000, 1000, 1000},
```
Comment (davd-gzl) [inline examples/gno.land/p/demo/tokens/grc721/grc721_royalty_test.gno:97]: Missing:
- Negative BPS
- Zero BPS
- Truncation edge case (salePrice=1, Bps=1000) ; Royalty truncate to 0 because `royaltyAmount` is a `int64`
- Large salePrice around int64 max -- should overflow or be protected by https://github.com/gnolang/gno/pull/5504/changes#r3088182613

Comment (notJoon) [reply examples/gno.land/p/demo/tokens/grc721/grc721_royalty_test.gno:97]: https://github.com/gnolang/gno/pull/5504/commits/87f8aeb25799145251c98173f951fa9fd2ed2ab4

File: examples/gno.land/p/demo/tokens/grc721/grc721_royalty.gno (hunk 2 above)
Comment (notJoon) [reply examples/gno.land/p/demo/tokens/grc721/grc721_royalty.gno]: https://github.com/gnolang/gno/pull/5504/commits/87f8aeb25799145251c98173f951fa9fd2ed2ab4

**Review by davd-gzl — APPROVED** [review]
